xds/internal/client/client_callback.go

```diff
@@ -34,6 +34,7 @@ func (c *Client) scheduleCallback(wi *watchInfo, update interface{}, err error)
 
 func (c *Client) callCallback(wiu *watcherInfoWithUpdate) {
 	c.mu.Lock()
+	wiu.wi.expiryTimer.Stop()
```

Comment:
Should we put this in `scheduleCallback` instead?

Reply:
This is the sequence of events I'm thinking about
- watcher registered at time = 0
- response received at `expiryTime - delta`
- we schedule a callback with the results from the response
- `delta` passed by
- the expiry timer is stopped and another callback is scheduled with an error saying watcher timeout

Eventually, application will receive both callbacks.

Reply:
That's right! And I also found that expiry timer doesn't need the mutex. I moved it to `scheduleCallback()`.

This reduces the window, but race can still happen when timeout races with the response. Because timer.Stop() doesn't ensure the callback won't happen (even with `if !timer.Stop { <- timer.C }`)

To really solve this problem, we need a boolean, and skip timeout error. I added a commit, PTAL.

Reply:
Hmm, I moved `timer.Stop` back to `callCallback()`, because race detector was complaining.

The race is between the afterFunc body, and assigning to expiryTimer. It could happen if timeout is very short, so func (which read timer variable) is executed at the same time the timer variable is assigned.
```go
	wi.expiryTimer = time.AfterFunc(defaultWatchExpiryTimeout, func() {
		c.scheduleCallback(wi, ClusterUpdate{}, errCallbackTimeout)
	})
```

Reply:
This is actually quite hard to follow now with things happening at multiple places. How about doing the following:
- add a `state` enum to the `watchInfo` struct with possible values being `{created, response_received, timer_expired, cancelled}`. And this `state` will be protected by a mutex.
- in the `AfterFunc`, you grab the mutex and check the `state` and if it is already in `response_received`, you do nothing. You dont even queue a callback.
- in the `newxDSUpdate` funcs, before queuing the callback, we set `state` to `response_received`, stop the timer and queue the callback. In this case we might invoke the callback with a good response after invoking it with a timeout error. If we strictly want to avoid that, we can check here if the `state` is already `timer_expired` and if so, do nothing.
- in the `cancel` returned by the watch API, we should set the state to `cancelled`.

I feel this might be cleaner. What do you think?

Reply:
Done. Also added methods so we remember to check the state before scheduling callback.